File: packages/pi-tai/src/config/schema.ts
import {
  DEFAULT_MODEL_PROFILES,
  type ModelProfile,
} from "../model-profiles/domain.ts";

export const TITLE_EFFORTS = [
  "minimal",
  "low",
  "medium",
  "high",
  "xhigh",
  "max",
] as const;

export type TitleEffort = (typeof TITLE_EFFORTS)[number];

export interface SessionTitleConfig {
  provider?: string;
  model?: string;
  effort: TitleEffort;
  maxWords: number;
  fallback: "heuristic";
}

export interface AnsiThemeConfig {
  darkTheme: string;
  lightTheme: string;
  pollIntervalMs: number;
}

export interface NotificationsConfig {
  reviewFailure: boolean;
  agentCompletion: boolean;
}

export interface CompactionConfig {
  enabled: boolean;
  thresholdPercent: number;
}

/**
 * Host-owned, per-session, command-mutable policy. Selects models and spends tokens, so it is
 * resolved by the authority that owns the session rather than by whichever client is attached.
 */
export interface SessionPolicy {
  sessionTitle: SessionTitleConfig;
  compaction: CompactionConfig;
  modelProfiles: readonly ModelProfile[];
}

/**
 * Client-local presentation preferences. These never cross the wire: an editor client owns its
 * own theming and decides for itself how to surface notifications.
 */
export interface ClientPreferences {
  ansiTheme: AnsiThemeConfig;
  notifications: NotificationsConfig;
}

/**
 * Machine-scoped, admin-writable configuration. Deliberately empty until checkpoint 12 supplies
 * Guardian's reviewer model and timeout, which are security-sensitive and must land separately.
 * `Record<string, never>` rather than an empty interface so it admits only `{}`.
 */
export type HostMachineConfig = Record<string, never>;

export interface ResolvedPiTaiConfig {
  sessionPolicy: SessionPolicy;
  clientPreferences: ClientPreferences;
  hostMachine: HostMachineConfig;
}

export const DEFAULT_SESSION_POLICY: SessionPolicy = Object.freeze({
  sessionTitle: Object.freeze({
    effort: "minimal",
    maxWords: 6,
    fallback: "heuristic",
  }),
  compaction: Object.freeze({
    enabled: true,
    thresholdPercent: 90,
  }),
  modelProfiles: DEFAULT_MODEL_PROFILES,
});

export const DEFAULT_CLIENT_PREFERENCES: ClientPreferences = Object.freeze({
  ansiTheme: Object.freeze({
    darkTheme: "ansi-dark",
    lightTheme: "ansi-light",
    pollIntervalMs: 2000,
  }),
  notifications: Object.freeze({
    reviewFailure: true,
    agentCompletion: true,
  }),
});

export const DEFAULT_HOST_MACHINE_CONFIG: HostMachineConfig = Object.freeze({});

export const DEFAULT_PI_TAI_CONFIG: ResolvedPiTaiConfig = Object.freeze({
  sessionPolicy: DEFAULT_SESSION_POLICY,
  clientPreferences: DEFAULT_CLIENT_PREFERENCES,
  hostMachine: DEFAULT_HOST_MACHINE_CONFIG,
});

File: crates/config/src/schema.rs
use serde::{Deserialize, Serialize};
use specta::Type;

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum ThinkingEffort {
    Off,
    Minimal,
    Low,
    Medium,
    High,
    Xhigh,
    Max,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CompactionConfig {
    pub enabled: bool,
    #[specta(type = i32)]
    pub threshold_percent: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ModelProfile {
    pub name: String,
    pub provider: String,
    pub model: String,
    pub effort: ThinkingEffort,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SessionPolicy {
    pub compaction: CompactionConfig,
    pub model_profiles: Vec<ModelProfile>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AnsiThemeConfig {
    pub dark_theme: String,
    pub light_theme: String,
    pub poll_interval_ms: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NotificationsConfig {
    pub review_failure: bool,
    pub agent_completion: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CmuxConfig {
    pub enabled: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ClientPreferences {
    pub ansi_theme: AnsiThemeConfig,
    pub notifications: NotificationsConfig,
    pub cmux: CmuxConfig,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq, Eq, Default)]
#[serde(rename_all = "camelCase")]
pub struct HostMachineConfig {}

#[derive(Debug, Clone, Serialize, Deserialize, Type, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ResolvedPiTaiConfig {
    pub session_policy: SessionPolicy,
    pub client_preferences: ClientPreferences,
    pub host_machine: HostMachineConfig,
}

pub fn default_config() -> ResolvedPiTaiConfig {
    ResolvedPiTaiConfig {
        session_policy: SessionPolicy {
            compaction: CompactionConfig {
                enabled: true,
                threshold_percent: 90.0,
            },
            model_profiles: ["low", "medium", "high"]
                .into_iter()
                .map(|effort| ModelProfile {
                    name: format!("sol-{effort}"),
                    provider: "openai-codex".into(),
                    model: "gpt-5.6-sol".into(),
                    effort: match effort {
                        "low" => ThinkingEffort::Low,
                        "medium" => ThinkingEffort::Medium,
                        _ => ThinkingEffort::High,
                    },
                })
                .collect(),
        },
        client_preferences: ClientPreferences {
            ansi_theme: AnsiThemeConfig {
                dark_theme: "ansi-dark".into(),
                light_theme: "ansi-light".into(),
                poll_interval_ms: 2000,
            },
            notifications: NotificationsConfig {
                review_failure: true,
                agent_completion: true,
            },
            cmux: CmuxConfig { enabled: true },
        },
        host_machine: HostMachineConfig {},
    }
}
